### ISSUE TITLE
Issue with Persistence Provider and Workflow Instance API

### ISSUE THREAD

**Issue by harshad-azilen** [elsa-workflows/elsa-core#2552]
Hi @sfmskywalker ,

Thanks for your reply in discord channel. I am putting issue details here is comment so that its visible and accessible quickly.

Background: We are using Mongo persistence provider for ELSA workflow and we are facing issue to show workflow instance on Dashboard. Earlier implementation using in-memory so all is working fine but after using persistence provider we are facing issue.

Analysis:- ELSA API storing Workflow definition and instance into database. Once we click on Workflow Instance tab on Dashboard then its calling API to get instance data from database. And in our case we are having issue.

We debug further and found it seems like issue with  SetVariable method. We are assigning order entity data to OrderDetail variable like **activityExecutionContext.SetVariable("OrderDetail", order).**

This variable saved along with WorkFlowInstance json under "variables". See snap. While retrieving instances,  ELSA api having issue. If i remove this variable from stored JSON then its working. 

![image](https://user-images.githubusercontent.com/95464910/144559999-11152399-9a2f-4a29-8852-ed2ab280cc71.png)

To resolve, could you please advice do we need to set mapper anywhere for this or its an issue with serialization/de-serialization or any other issue?

**Comment by sfmskywalker** [comment]
Hi @harshad-azilen 

What is the exact issue you are seeing? Do you see an error, and if yes, what is the error?

**Comment by sfmskywalker** [comment]
Presuming it has something to do with deserialization of the document, you could try registering your custom type with `BsonClassMap`.

For example, the Mongo DB provider for Elsa uses [the following code](https://github.com/elsa-workflows/elsa-core/blob/master/src/persistence/Elsa.Persistence.MongoDb/Services/DatabaseRegister.cs#L49) to create class maps:

```csharp
BsonClassMap.RegisterClassMap<Bookmark>(cm => cm.AutoMap());
BsonClassMap.RegisterClassMap<WorkflowExecutionLogRecord>(cm => cm.AutoMap());
BsonClassMap.RegisterClassMap<WorkflowOutputReference>(cm => cm.AutoMap());
```

**Comment by harshad-azilen** [comment]
Hi @sfmskywalker , no any error except Dashboard error

![image](https://user-images.githubusercontent.com/95464910/144580395-62713186-e1f0-49b7-82bb-d6ca89008014.png)

**Comment by harshad-azilen** [comment]
> Presuming it has something to do with deserialization of the document, you could try registering your custom type with `BsonClassMap`.
> 
> For example, the Mongo DB provider for Elsa uses [the following code](https://github.com/elsa-workflows/elsa-core/blob/master/src/persistence/Elsa.Persistence.MongoDb/Services/DatabaseRegister.cs#L49) to create class maps:
> 
> ```cs
> BsonClassMap.RegisterClassMap<Bookmark>(cm => cm.AutoMap());
> BsonClassMap.RegisterClassMap<WorkflowExecutionLogRecord>(cm => cm.AutoMap());
> BsonClassMap.RegisterClassMap<WorkflowOutputReference>(cm => cm.AutoMap());
> ```

Hi @sfmskywalker , could you please advice how to do same for PostgreSQL?

Also do we need to add this in to startup.cs file or any other location?

**Comment by sfmskywalker** [comment]
> no any error except Dashboard error

I can't reproduce this behavior with the latest version of Elsa. Can you try with the 1.4.x preview packages available from MyGet? Odds are that the issue you're seeing has already been resolved.

If you still do see this issue after upgrading, then please send me a repro that I can look at.

> could you please advice how to do same for PostgreSQL?
You shouldn't have to do this for PostgreSQL, since the EF Core provider handles mapping itself using Newtonsoft Json.
It's only necessary for the MongoDB provider, which relies on MongoDB's serialization implementation.

> Also do we need to add this in to startup.cs file or any other location?
For Mongo, the place to do it would indeed be Startup.

**Comment by harshad-azilen** [comment]
Hi @sfmskywalker , just to update you that I used PostgreSQL to double check and its also having same issue. 

I think its an issue with custom type when we assign to variable in workflow. If we assign values like string, int then no such issue exists but if we set custom type like class then issues exists.

Due to some dependency and complexity its difficult to share code/repo but yes i can share export file of workflow and few details below. 

I have simplified flow, see snap.  

1. Start with POST /initiate and pass body in parameter
2. setting value to orderid variable using input.Body.OrderId 
3. checking orderid is null or not using isNullOrEmpty(OrderId) 
4. Set Variable in workflow using SetOrderDetail(OrderId) here SetOrderDetail method exits in code.

![image](https://user-images.githubusercontent.com/95464910/144600582-60eb6a63-8c62-4058-895f-d69ded3c937c.png)

[test-work-flow.txt](https://github.com/elsa-workflows/elsa-core/files/7649642/test-work-flow.txt)

**Comment by harshad-azilen** [comment]
> I can't reproduce this behavior with the latest version of Elsa. Can you try with the 1.4.x preview packages available from MyGet? Odds are that the issue you're seeing has already been resolved.

Hi @sfmskywalker , i am facing issue with n **Jint** package while updating all packages to 2.4.x preview. See below project reference and error snap.

![image](https://user-images.githubusercontent.com/95464910/144725281-f4c5b882-f342-4783-bc5d-d78b569db300.png)

**Comment by sfmskywalker** [comment]
Elsa is currently using latest preview builds from the prerelease Jint feed. Updating your NuGet.config file with that feed will fix it.

https://www.myget.org/feed/jint/package/nuget/Jint

**Comment by harshad-azilen** [comment]
> Elsa is currently using latest preview builds from the prerelease Jint feed. Updating your NuGet.config file with that feed will fix it.
> 
> https://www.myget.org/feed/jint/package/nuget/Jint

Thanks for your reply. I am able to refer Jint package now. After updating all packages to latest **still i am facing same issue**. See snap below.

![image](https://user-images.githubusercontent.com/95464910/144726381-512bb45c-11fa-43b1-8eda-58ef20644f21.png).

I am using PostgreSQL. I suspect issue is just because of custom type which was persist along with other data. See attached file contains json object which was stored in table "Elsa"."WorkflowInstances">> Column Data.

[workflowinstance-table-data.txt](https://github.com/elsa-workflows/elsa-core/files/7654834/workflowinstance-table-data.txt)

If we remove " "$type": "Shared.BGC.Entities.OrderManagement.Order, Shared.BGC"," from stored data and reload page then its working.

![image](https://user-images.githubusercontent.com/95464910/144726567-2b43a0b5-0ae1-4ac8-9e90-494e1921ad80.png)

**Comment by harshad-azilen** [comment]
> Elsa is currently using latest preview builds from the prerelease Jint feed. Updating your NuGet.config file with that feed will fix it.
> 
> https://www.myget.org/feed/jint/package/nuget/Jint

Hi @sfmskywalker , there is some issue with Jint feed. Any idea?  I am using below command.

Install-Package Jint -Version 1482256692.0.0 -Source https://www.myget.org/F/jint/api/v3/index.json 

See below error.

Install-Package : Failed to retrieve information about 'Jint' from remote source 'https://www.myget.org/feed/jint/package/nuget/Jint/FindPackagesById()?id='Jint'&semVerLevel=2.0.0'.
  The content at 'https://www.myget.org/feed/jint/package/nuget/Jint/FindPackagesById()?id='Jint'&semVerLevel=2.0.0' is not valid XML.
  For security reasons DTD is prohibited in this XML document. To enable DTD processing set the DtdProcessing property on XmlReaderSettings to Parse and pass the settings into XmlReader.Create 
method.
At line:1 char:1
+ Install-Package Jint -Version 1482256692.0.0 -Source https://www.myge ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
 
Time Elapsed: 00:00:03.9304978
PM>

**Comment by harshad-azilen** [comment]
Hi @sfmskywalker , request to have a look at below root cause for issue and do needful for solution.

After troubleshooting a lot, we are able to find root cause with both provider. See below details and let us know your input.

**Mongo Provider:- There is issue with de serializing.** You can check below errors and its status.

1. Unknown discriminator value 'FinishOutput'- **Fix** with BsonClassMap.RegisterClassMap<FinishOutput>(cm => cm.AutoMap());
2. Unknown discriminator value HttpRequestModel- **FIx** with BsonClassMap.RegisterClassMap<HttpRequestModel>(cm => cm.AutoMap());
3. An error occurred while deserializing the Type property of class Elsa.Models.SimpleException: Specified method is not supported. - **Not fix.** Let us know any solution. As ELSA api storing Type value in document and its causing issue

**PostgreSQL:- There is issue with de-serializing.** You can check below errors and its status.

1. Error converting value "61af0395ec2bca5722a2dd97" to type 'MongoDB.Bson.ObjectId'.  Path 'variables.data.OrderDetail.applicantInfo.id', line 222, position 42.- **Not Fix**. We have ApplicantInfo class having property Id as ObjectId . Here ELSA api storing json object into database and while reading data its causing issue. 

Could you please let us know solution for points having status as  **Not Fix**.

**Comment by sfmskywalker** [comment]
Hi @harshad-azilen

If you provide steps to reproduce the issues you reported then I'd be happy to look into it.
The easiest way for me to troubleshoot is if I get a (stripped down) solution that demonstrates the issues you mentioned.

**Comment by harshad-azilen** [comment]
> Hi @harshad-azilen
> 
> If you provide steps to reproduce the issues you reported then I'd be happy to look into it. The easiest way for me to troubleshoot is if I get a (stripped down) solution that demonstrates the issues you mentioned.

Hi @sfmskywalker ,  thanks for reply and support. Issue was resolved and we can close this ticket now.

Root cause: ELSA packages not able to do proper de-serializing of stored object

Resolution: 
- Issue was resolved with ELSA myget packages version 2.4.x preview
- referred Jint package
- Created ObjectId converter to solve de-serializing issue. We have model property with type ObjectId. and once workflow-instance API try to read records from WorkflowInstance collection then this was causing issue.

**Comment by sfmskywalker** [comment]
Awesome, thanks for letting us know the resolution 👍🏻
